### PR TITLE
Implement Mediation policies

### PR DESCRIPTION
**Pull request by thivindu**
This pull request introduces new API Gateway policy definitions. The main focus is on expanding the set of default policies available for API request/response manipulation.

**New Policy Definitions:**

* Added new policy: `add-headers` — allows adding arbitrary headers to requests and/or responses without overwriting existing values.
* Added new policy: `remove-headers` — enables removing specified headers from the requests and/or responses, with proper handling of non-existing headers.
* Added new policy: `set-headers` — enables setting specified headers to the requests and/or responses.
* Added new policy: `json-to-xml` — provides the ability to transform JSON payloads to XML in both request and response flows, with safety checks for payload type.
* Added new policy: `xml-to-json` — provides the ability to transform XML payloads to JSON in both request and response flows, with safety checks for payload type.
* Added new policy: `log-message` — provides the ability to log the payload and headers of request/response messages.

## Summary by CodeRabbit

* **New Features**
  * Six new policies: Add-Headers, Set-Headers, Remove-Headers, Log-Message, JSON-to-XML, XML-to-JSON — configurable request/response header and payload handling and transformations.

* **Tests**
  * Comprehensive test suites added covering validation, request/response flows, edge cases and conversions for each new policy.

* **Chores**
  * .gitignore updated to ignore IDE/project metadata and macOS files.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: policies/add-headers/go.mod [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module github.com/policy-engine/policies/add-headers
+
+go 1.25.1
```
[the remark this answers is not in the document]
Comment (thivindu) [reply policies/add-headers/go.mod:3]: This is not an issue and all other policies use this.

File: policies/json-to-xml/jsontoxml_test.go [hunk 2]
```diff
@@ -0,0 +1,1399 @@
...
+
+	xmlBytes, err := p.ConvertJSONToXML(jsonData)
+	if err != nil {
+		t.Fatalf("Expected no error, got: %v", err)
+	}
+
+	xmlStr := string(xmlBytes)
+	if !strings.Contains(xmlStr, "<root>") && !strings.Contains(xmlStr, "</root>") {
+		t.Errorf("Expected root element for empty object")
+	}
```
[the remark this answers is not in the document]
Comment (thivindu) [reply policies/json-to-xml/jsontoxml_test.go:614]: Fixed via - https://github.com/wso2/gateway-controllers/pull/4/commits/6875e64ab0e3a24e4dd36da8646a19a3a991b11d

File: policies/log-message/logmessage.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+			return
+		}
+
+		// Add header to map
+		if len(values) == 1 {
+			headersMap[name] = values[0]
+		} else {
+			headersMap[name] = values
+		}
+	})
```
[the remark this answers is not in the document]
Comment (thivindu) [reply policies/log-message/logmessage.go:185]: This is by design to conform with previous implementation of this policy

File: policies/set-headers/setheaders.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		entry := HeaderEntry{
+			Name:  strings.ToLower(strings.TrimSpace(headerMap["name"].(string))), // Normalize to lowercase
+			Value: headerMap["value"].(string),
+		}
+
+		entries = append(entries, entry)
+	}
+
+	return entries
+}
```
[the remark this answers is not in the document]
Comment (thivindu) [reply policies/set-headers/setheaders.go:149]: This would be validated against the policy definition when the API is created with the policy attached. Therefore there would be no runtime panics.

File: policies/log-message/logmessage.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+	return headersMap
+}
```
[the remark this answers is not in the document]
Comment (thivindu) [reply policies/log-message/logmessage.go:188]: The `(h *Headers) Get(name string) []string` and `(h *Headers) Iterate(fn func(name string, values []string))` check if the headers are nil and if it is they return. nil or empty. So there would be no runtime panics.

https://github.com/wso2/api-platform/blob/3909ea060aab0394de338fcec275ccaa16edc604/sdk/gateway/policy/v1alpha/headers.go#L29-L39

https://github.com/wso2/api-platform/blob/3909ea060aab0394de338fcec275ccaa16edc604/sdk/gateway/policy/v1alpha/headers.go#L89-L97

File: policies/json-to-xml/jsontoxml.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+				for _, item := range arr {
+					childElement := p.buildXMLStruct(item, key)
+					element.Children = append(element.Children, childElement)
+				}
+			} else {
+				// For non-arrays, process normally
+				childElement := p.buildXMLStruct(value, key)
+				element.Children = append(element.Children, childElement)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (thivindu) [reply policies/json-to-xml/jsontoxml.go:230]: Addressed via - https://github.com/wso2/gateway-controllers/pull/4/changes/e7ae4d31df00c95abdf4c666d82b46615ab4e4cf

File: policies/json-to-xml/jsontoxml.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	}
+
+	// Check content type to ensure it is JSON
+	contentType := ""
+	if contentTypeHeaders := ctx.Headers.Get("content-type"); len(contentTypeHeaders) > 0 {
+		contentType = strings.ToLower(contentTypeHeaders[0])
+	}
+
+	if !strings.Contains(contentType, "application/json") {
+		return p.handleInternalServerError("Content-Type must be application/json for JSON to XML transformation")
```
Comment (renuka-fernando) [inline policies/json-to-xml/jsontoxml.go:76]: This shouldn't be a server error, it should be a 400.

I think, if the content is already XML, we can pass it to the upstream and otherwise 400
@malinthaprasan wdyt? I am not sure whether our existing Bijira policy works.

Comment (thivindu) [reply policies/json-to-xml/jsontoxml.go:76]: In Bijira the policy description says -

```This policy assumes that the request/response payload is XML. Attempting to use it on a request/response with a non-XML payload will result in premature termination of the mediation flow. For the same reason, this policy cannot be attached multiple times to a resource since once it is used, the payload will be a JSON value.```

And in the code 

```
xml xmlPayload = check req.getXmlPayload();
```

will fail and return from the function immediately.

And it will return a 500 Internal server error.

**Review by Krishanx92 — APPROVED** [review]
